src/layout.rs

```diff
@@ -366,161 +366,147 @@ impl Layout {
     }
+    // ensure the first element touches the left/top edge of the area
     if let Some(first) = elements.first() {
-        ccs.push(match layout.direction {
-            Direction::Horizontal => first.left() | EQ(REQUIRED) | f64::from(dest_area.left()),
-            Direction::Vertical => first.top() | EQ(REQUIRED) | f64::from(dest_area.top()),
-        });
+        solver.add_constraint(first.start | EQ(REQUIRED) | area.start)?;
     }
+    // ensure the last element touches the right/bottom edge of the area
     if layout.expand_to_fill {
         if let Some(last) = elements.last() {
-            ccs.push(match layout.direction {
-                Direction::Horizontal => last.right() | EQ(REQUIRED) | f64::from(dest_area.right()),
-                Direction::Vertical => last.bottom() | EQ(REQUIRED) | f64::from(dest_area.bottom()),
-            });
+            solver.add_constraint(last.end | EQ(REQUIRED) | area.end)?;
         }
     }
-    match layout.direction {
-        Direction::Horizontal => {
-            for pair in elements.windows(2) {
-                ccs.push((pair[0].x + pair[0].width) | EQ(REQUIRED) | pair[1].x);
+    // apply the constraints
+    for (&constraint, &element) in layout.constraints.iter().zip(elements.iter()) {
+        match constraint {
+            Constraint::Percentage(p) => {
+                let percent = f64::from(p) / 100.00;
+                solver.add_constraint(element.size() | EQ(STRONG) | (area.size() * percent))?;
             }
-            for (i, size) in layout.constraints.iter().enumerate() {
-                ccs.push(elements[i].y | EQ(REQUIRED) | f64::from(dest_area.y));
-                ccs.push(elements[i].height | EQ(REQUIRED) | f64::from(dest_area.height));
-                ccs.push(match *size {
-                    Constraint::Length(v) => elements[i].width | EQ(MEDIUM) | f64::from(v),
-                    Constraint::Percentage(v) => {
-                        elements[i].width | EQ(MEDIUM) | (f64::from(v * dest_area.width) / 100.0)
-                    }
-                    Constraint::Ratio(n, d) => {
-                        elements[i].width
-                            | EQ(MEDIUM)
-                            | (f64::from(dest_area.width) * f64::from(n) / f64::from(d))
-                    }
-                    Constraint::Min(v) => elements[i].width | GE(MEDIUM) | f64::from(v),
-                    Constraint::Max(v) => elements[i].width | LE(MEDIUM) | f64::from(v),
-                });
-
-                match *size {
-                    Constraint::Min(v) | Constraint::Max(v) => {
-                        ccs.push(elements[i].width | EQ(WEAK) | f64::from(v));
-                    }
-                    _ => {}
-                }
+            Constraint::Ratio(n, d) => {
+                // avoid division by zero by using 1 when denominator is 0
+                let ratio = f64::from(n) / f64::from(d.max(1));
```

Comment:
I think if `d` is zero, `d` should be `n`, i.e. ratio == 1. Right now, if you have ratio `Ratio(10, 0)` a strong constraint will be added that will try to make the rect size 10 times the target area and that constraint is always going to be violated.

Reply:
I think I'm fine with leaving this for now as it matches the behavior in `Constraint::apply()`. Current behavior is to panic, which is probably more correct.